### DISCUSSION TITLE
How to stabilize SpotBugs release with rapid feature enhancement

### DISCUSSION THREAD

**Discussion by KengoTODA** [spotbugs/spotbugs#2076] [General]
SpotBugs welcomed many PRs luckily. For instance, we merged [23 PRs (excluding PRs from dependabot) from 2022-01-01 to 2022-05-29](https://github.com/spotbugs/spotbugs/pulls?q=is%3Apr+is%3Amerged+created%3A%3E%3D2022-01-01+-label%3Adependencies). This is a really good signal proving the activeness of the SpotBugs community, I believe. Thanks for your continuous support!

Today I want to discuss with you a problem.

## The problem to focus on

This activeness of the SpotBugs project brings a new problem. Recently we got several issues regarding new detectors, and users have less opportunity to have stable SpotBugs because we have had fewer patch releases recently. Let's have numbers to hold a constructive discussion:

* We had 3 major/minor releases and 10 patch releases in 2020.
* We had 3 minor releases and 7 patch releases in 2021. But,
* In 2022 we have already released 2 minor releases and only 1 patch release during these 5 months.

The minor-release frequency is 60% more than 2020-2021 (comparing 2/5 to 6/24).

Thanks for JUnit and [SpotBugs Test Harness](https://github.com/spotbugs/spotbugs/tree/68e7dca67c782c794518b9d2ded4dd32e044dac3/test-harness), we could find many defects during the PR review phase. However, they are not enough to catch problems caused in various languages/platforms/libraries.

## Ideas

Currently, I have a few ideas to make the situation better:

0. Request users to run the analysis with the latest SNAPSHOT version continuously, to find problems before the official release.
1. Release RC (release candidate) versions before we release new minor updates.
    * Just like what we did at the 4.0 launch.
2. Pick some repositories to run the community testing, and run during the PR review phase.
    * Better to pick projects not only from the Java community but also from other JVM languages&apos; one like Kotlin.
    * Not sure suitable projects to pick. Any suggestions? Hope that the project can be built successfully with `./gradlew spotbugsMain` without additional configuration.

There are more ideas as follows, but for now, I personally have no motivation to introduce them due to the lack of manpower from the SpotBugs team (please try to understand that SpotBugs is not a project backed by organizations):

0. Maintain not only the latest minor version but also the previous minor version and backport bug fixes onto this version.
    * Actually, SpotBugs did this previously; we had `master` and `release-3.1` branches, and I gave up doing so due to too heavy maintenance cost.
1. Have another release train for LTS (long-term support) version.
    * This is more difficult than the previous one.
2. Do not merge feature enhancement so frequently, and merge them in batch, like once per quarter.
    * It just demotivates contributors, so I hope that we try another way first.

That's all from my side, please share your thought and opinions. Thanks!

**Comment by ThrawnCA** [comment]
RC versions seem like a low-overhead option.

**Comment by iloveeclipse** [comment]
I don't see the problem. If we release more often, people have possibility, to get fixes faster. Those who see regressions may want to either contribute a fix or stay on older versions, no one forced them to move to the latest greatest.

At the end it's all about resources and that is what we don't have, independently which solution is proposed.

**Comment by gtoison** [comment]
One way to address this is to raise the bar for contributions (especially for new detector) on the unit tests front.
Checking for loops? Your unit test needs to check for `while`, `do while`,`break`,`continue` and `for`. 
Checking for method signatures? You need to have some interfaces, abstract classes, enums, records, final and static methods, etc.

Maybe these examples don't make sense (and maybe they would have not detected the recent issues) but the idea is to have a checklist of what needs to be tested. The unit tests tend to focus on the bugs that the contributors want to detect, but it's harder to think about what other people might code.
When an issue is found the checklist can be amended.

IMO there are two type of issues:
- False positives/negatives: these are not easy to spot with automated testing but the impact for user is not too bad. If the releases get more frequent theses issues will crop up all at once and it will be harder to locate the cause, because so much code has changed. Running SpotBugs on other projects would require some efforts: someone will need to sort out the new/resolved bugs.
- Crashes: these could be spotted automatically by running SpotBugs on some non-trivial open source projects.

**Reply by gtoison** [reply]
Side question: I've noticed that the last analysis of the project by SonarCloud was from 2 years ago: https://sonarcloud.io/summary/new_code?id=com.github.spotbugs.spotbugs
As a result the code coverage badge of the project's [README](https://github.com/spotbugs/spotbugs/blob/master/README.md) is also from two years ago.

Was SonarCloud abandoned because it caused problems? Or it stopped working and nobody was really using it so it wasn't fixed?

**Reply by KengoTODA** [reply]
Seems that the [PR analysis is working](https://sonarcloud.io/project/pull_requests_list?id=com.github.spotbugs.spotbugs) but branch analysis is not. I will check.

**Reply by KengoTODA** [reply]
https://github.com/spotbugs/spotbugs/pull/2105

**Comment by baloghadamsoftware** [comment]
I would also vote on the RC versions.

Furthermore, on my side I promiss that I will always execute our own test suite (based on open-source projects) on every PR of our student developers before they are accepted. I will allso post the results so you can see yourself the impact of their work. Actually, I am already doing it as you can see at some PRs.

**Comment by baloghadamsoftware** [comment]
Also, I would propose to have some extended CI test for each PR, similair what FindSecBugs has in their GitHub CI. As an alternative, this is also worth to consider: #1746. Of course, it should be extended with other open-source projects.

Having RC versions and extended test suites would also increase confidence of the whole community and make it easier for large infrastructure PRs to get accepted, such as e.g. #2073.

**Comment by jtnord** [comment]
Could alternative approach could be to make newly introduced detectors go through an incubating phase where a user needs to opt into incubating detectors (using a flag or such in `spotbugs-maven-plugin` or the like?
This allows users to still upgrade regularly, to get fixes without the overhead of a stable line.  And new detectors can graduate when after a set bar (some time with no false positives reported from users that are using the incubating matcher?)

I have no idea if this even possible - I am just a regular user.

This does beg the question how many people would run incubating features - but that would also be the same for RCs.
